### PR TITLE
Rework selectors to support selection by text and n-th element selection (closes #615)

### LINKED ISSUE #615: Rework selectors to support selection by text and n-th element selection
Seems like we will need to use a little bit different machinery for this one to make it work in raw API (dedicated factory and executor)

**Update**
*Final API* (see discussion below):
`Selector(fn, options)`
`ClientFunction(fn, options)`
`t.select(fn, options)`
`t.eval(fn, options)`
`selector.with(options)`
`clientFunction.with(options)`

`options` are:
`dependencies` - functions and variables used by selector code 
`boundTestRun` - bound test run

Selector-specific `options`:
`visibilityCheck` - wait for element to become visible
`timeout` - element waiting timeout
`textFilter` - regexp or string, filters returned element or element collection by specified text
`index` - if selector returns enumerable collection specifies index of the selected element in that collection, 0 by default. If node returned `index` is ignored.

-----

- [x] Rename client function factories to builders
- [x] Pass `options` to ctor
- [x] `.with()` should merge options and create new builder and function
- [x] Rename `scopeVars` to `dependencies`
- [x] Implement `index`, let selector return DOM node, DOM collection or array. throw otherwise
- [x] Implement `textFilter`

### PR DESCRIPTION
**Pull request by inikulin**
\cc @AlexanderMoskovkin @AndreyBelym @VasilyStrelyaev

### CODE REVIEW

File: src/errors/test-run/templates.js [hunk 1]
```diff
@@ -234,8 +234,8 @@ export default {
        ${err.getCallsiteMarkup()}
     `),
 
-    [TYPE.nonDomNodeSelectorResultError]: err => markup(err, `
-        ${err.instantiationCallsiteName} can only return a DOM node, null or undefined. Use ClientFunction to return other values.
+    [TYPE.invalidSelectorResultError]: err => markup(err, `
+        Function that specifies a selector can only return a DOM node, an array of nodes, NodeList or HTMLCollection, null or undefined. Use ClientFunction to return other values.
```
Comment (VasilyStrelyaev) [inline src/errors/test-run/templates.js]: I must've overthought with the enumeration `NodeList or HTMLCollection, null or undefined`. Let's put it simple
```
        Function that specifies a selector can only return a DOM node, an array of nodes, NodeList, HTMLCollection, null or undefined. Use ClientFunction to return other values.
```

**Comment by VasilyStrelyaev** [comment]
\r-

**Comment by inikulin** [comment]
@testcafe-build-bot \retest

**Comment by inikulin** [comment]
ping @AlexanderMoskovkin @VasilyStrelyaev @AndreyBelym

**Comment by AndreyBelym** [comment]
lgtm

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Comment by VasilyStrelyaev** [comment]
lgtm

**Pull request merged**